Fix #539 "cast not available"

GregoireDruant opened the pull request:
Closes #539 "cast not available" message when viewing tvshow or episode info in kodi. 
Works with add-on mode and native mode as well.

The fix only consist in setting the mediatype and scrapper for each path entry in kodi database.
Normally only parent paths have this information, but jellyfin-kodi does not set parentPathIds.
Also I tryed setting the parentPathIds but this does not fix the "cast not available" message if the parent path is not a regular folder (does not work with plugin://).

oddstr13 commented:
Apologies, I don't have time to review this for a while.

On 6 May 2023 10.57.36 CEST, GregoireDruant ***@***.***> wrote:
>Fixes the &quot;cast not available&quot; message when viewing tvshow or episode info in kodi, issue #539 
>Works with add-on mode and native mode as well.
>
>The fix only consist in setting the mediatype and scrapper for each path entry in kodi database.
>Normally only parent paths have this information, but jellyfin-kodi does not set parentPathIds.
>Also I tryed setting the parentPathIds but this does not fix the &quot;cast not available&quot; message if the parent path is not a regular folder (does not work with plugin://).
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/jellyfin/jellyfin-kodi/pull/731
>
>-- Commit Summary --
>
>  * Feat: Set mediaType and scrapper to all paths for movies and tvshows
>  * Merge branch &#39;fix/cast-not-available-blunt-method&#39;
>
>-- File Changes --
>
>    M jellyfin_kodi/objects/kodi/queries.py (7)
>
>-- Patch Links --
>
>https://github.com/jellyfin/jellyfin-kodi/pull/731.patch
>https://github.com/jellyfin/jellyfin-kodi/pull/731.diff
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/jellyfin/jellyfin-kodi/pull/731
>You are receiving this because you are subscribed to this thread.
>
>Message ID: ***@***.***&gt;

GregoireDruant commented:
Sorry there is an issue with this fix, regarding the information pannel for a tvshow episode.
If you put the cursor on an episode item, long press enter and chose information, it works.
However the information panel won't show up in these cases:
 - click on the episode when the default action is show information
 - if the default action on video item is "chose", chosing information from the menu that shows up when clicking on an episode
 - on a pc, type "i" when the cursor is on an episode.

If I remove the value "metadata.local" from the scrapper column, the info panel works again but without cast information.

I know how to fix this in native mode but not sure I can fix it in add-on mode.

Anyway I will commit a proper fix for native mode, that does not break anything in add-on mode.

I will do so by setting the path trail from episode to top parent path and set the mediatype and scrapper to the top parent path only, as kodi does.

GregoireDruant commented:
This is the proper fix, works perfectly in native mode, but does not fix add-on mode.

GregoireDruant commented:
After a lot of time spent to search how to make every kodi menu work in addon mode - for tvshows, which are the only videos that was not working ok - , the only way I cloud achieve it is by creating a top path entry with strPath = "plugin://plugin.video.jellyfin/" and strContent = tvshows, scrapper = metadata.local.

I found in the commits history that it used to be done like that but it was changed by Matt in in 2019, when he added the libraryId in the paths. I could not figure out why.

I did not remove his changes but I re added the top path with the plugin address and it seems to be working fine now.

So this PR includes fixes for both the add-on mode and native mode.

GregoireDruant commented:
I have been using this for 3 weeks both in native mode on a nvidia shield and in add-on mode on a windows pc, it has been working without issue.

Beside @oddstr13, can anyone review this PR ?

GregoireDruant commented:
I just added two new bugfixes to the music video library.
These are not related to bug 539, but since I already have this PR pending I figure I might as well commit here.

- Artist link to a music video was inserted using "insert or update", passing values that did not constitute an unique key, resulting in duplicate links (and appearing like duplicate videos in db, at least with my kodi settings).
- In musicvideos.py, obj['LibraryId'] and obj['LibraryName'] was incorrectly set, when they were correctly set a few lines before.
I removed the problematic lines of code (see diff).

oddstr13 commented:
I (and the Jellyfin project) has a strong preference for smaller, easier to review PRs, so please create another PR for the unrelated changes :slightly_smiling_face: 

I'm done with school for the summer, so I'll hopefully have more time for this project going forward

GregoireDruant commented:
> I (and the Jellyfin project) has a strong preference for smaller, easier to review PRs, so please create another PR for the unrelated changes 🙂
> 
> I'm done with school for the summer, so I'll hopefully have more time for this project going forward

I just reverted these commits and created PR #742

mcarlton00 commented on jellyfin_kodi/database/jellyfin_db.py:
```diff
@@ -71,6 +71,15 @@ def get_item_by_kodi_id(self, *args):
+
+            return self.cursor.fetchone()[0]
+        except TypeError:
+            return
```
I agree with the github scanner about not mixing implicit and explicit returns.  I believe in this case it's better to `return 0` (or `None` explicitly, but imo keeping the same type is generally better practice when feasible) and check the return value before attempting to use it on the calling function.

oddstr13 replied:
I was thinking the same, but this is the code style currently used in these files, so I don't consider it a blocker, cleaner to just fix them all in one go later some time.

mcarlton00 commented on jellyfin_kodi/objects/tvshows.py:
```diff
@@ -377,6 +398,12 @@ def episode_add(self, obj):
         self.jellyfin_db.add_reference(*values(obj, QUEM.add_reference_episode_obj))
+
+        parentPathId = self.jellyfin_db.get_episode_kodi_parent_path_id(*values(obj, QUEM.get_episode_kodi_parent_path_id_obj))
+        if obj['PathId'] != parentPathId:
```
```suggestion
        if parentPathId and obj['PathId'] != parentPathId:
```

Along with the return change from the queries, ensure the path actually exists before trying to use it.  It _shouldn't_ be an issue currently since the implicit return is `None`, but just an extra layer of security in case of weirdness happening.  And I would say necessary if we `return 0` instead.

oddstr13 commented:
@mcarlton00 said in chat that this PR functionally is good to go, and I think the styling lints are fine as is, considering the fact that that's how the old code base has been doing it, and as such is present all over these files.
